File: ts/src/day11.ts
import { DayBase } from './day-runner.js';

enum State {
    Resting = 0,
    Flashed = 1,
}

interface Octopus {
    power: number;
    state: State;
}

export class Day11 extends DayBase {
    constructor() {
        super(11);
    }

    public async run(): Promise<void> {
        const lines = await this.getInputLines();
        // const lines= [
        //     '5483143223',
        //     '2745854711',
        //     '5264556173',
        //     '6141336146',
        //     '6357385478',
        //     '4167524645',
        //     '2176841721',
        //     '6882881134',
        //     '4846848554',
        //     '5283751526',
        // ];

        let creatures: Octopus[][] = [];
        lines.forEach((line) => {
            let row: Octopus[] = [];
            for (let i = 0; i < line.length; i++) {
                let oc: Octopus = {
                    power: parseInt(line[i], 10),
                    state: State.Resting,
                };
                row.push(oc);
            }
            creatures.push(row);
        });

        let part1 = 0, part2 = 0;
        let rounds = 0, flashes= 0;
        while (part1 === 0 || part2 === 0) {
            rounds++;
            creatures.forEach(r => r.forEach(c => c.power += 1 ));

            let flasher = true;
            while (flasher) {
                flasher = false;
                for (let y = 0; y < creatures.length; y++) {
                    for (let x = 0; x < creatures[0].length; x++) {

                        let oc = creatures[y][x];
                        if (oc.power > 9 && oc.state == State.Resting) {
                            oc.state = State.Flashed;
                            flashes++;
                            flasher = true;
                            let neigh = this.getNeighbours(creatures, x, y);
                            neigh.forEach(n => {
                                n.power += 1;
                            });
                        }
                    }
                }
            }

            let allFlashed = true;
            creatures.forEach(r => r.forEach(c => {
                if (c.power > 9) {
                    c.power = 0;
                    c.state = State.Resting;
                }
                if (c.power != 0) { // if all in sync then they'll all be at 0 after a group flash :-)
                    allFlashed = false;
                }
            }));

            if (rounds == 100) {
                part1 = flashes;
            }
            if (allFlashed === true) {
                part2 = rounds;
            }
        }

        console.info(`Day ${this.day}`);
        console.info(`  Part 1:`, part1);
        console.info(`  Part 2:`, part2);
    }

    private getNeighbours(map: Octopus[][], x: number, y: number): Octopus[] {
        let result: Octopus[] = [];
        const www = map[0].length;
        const hhh = map.length;
        const delta = [[1,0], [1,1], [0,1], [-1,1], [-1,0], [-1,-1], [0,-1], [1,-1]];
        for (let [dx,dy] of delta) {
            if ((x+dx >= 0) && (x+dx < www) && (y+dy >= 0) && (y+dy < hhh)) {
                result.push(map[y+dy][x+dx]);
            }
        }
        return result;
    }
}

File: py/day11.py
# https://adventofcode.com/2021/day/11

with open("../input/day11-input", "r") as f:
    input = [line.strip() for line in f.readlines()]

seaO = []
for line in input:
    seaO.append([[int(c),False] for c in line])

rlen = len(seaO[0]) # assume square
sea = seaO.copy()
p1steps = 100
flashes, step = 0, 0
while True:
    step += 1
    # inc power
    for r in sea:
        for c in r:
            c[0] += 1
    # flash
    flashed = True
    while flashed:
        flashed = False
        for y,r in enumerate(sea):
            for x,c in enumerate(r):
                if c[0] > 9 and not c[1]:
                    for dx,dy in [(1,0),(1,1),(0,1),(-1,1),(-1,0),(-1,-1),(0,-1),(1,-1)]:
                        if x+dx >= 0 and x+dx < rlen and y+dy >= 0 and y+dy < rlen:
                            n = sea[y+dy][x+dx]
                            n[0] += 1
                    c[1] = True
                    flashed = True
                    flashes += 1
    # reset
    allZero = True
    for r in sea:
        for c in r:
            if c[0] > 9:
                c[0] = 0
                c[1] = False
            if c[0] > 0:  # part 2 check
                allZero = False

    if step == p1steps:
        part1 = flashes

    if allZero:  # everyone has just flashed
        part2 = step
        break

print("Day 11")
print("  Part 1:", part1)
print("  Part 2:", part2)
